src/profile-logic/call-node-info.js

```diff
@@ -331,21 +783,528 @@ export class CallNodeInfoInvertedImpl
-      (comparedCallNodeIndex) =>
-        compareNonInvertedCallNodesInSuffixOrderWithPath(
-          comparedCallNodeIndex,
-          callPath,
-          this._nonInvertedCallNodeTable
-        )
+    if (nodeHandle < this._rootCount) {
+      // nodeHandle is a root. For roots, the node handle IS the func index.
+      const funcIndex = nodeHandle;
+      const rangeStart =
+        funcIndex === 0
+          ? 0
+          : this._invertedRootCallNodeTable.suffixOrderIndexRangeEnd[
+              funcIndex - 1
+            ];
+      const rangeEnd =
+        this._invertedRootCallNodeTable.suffixOrderIndexRangeEnd[funcIndex];
+      return [rangeStart, rangeEnd];
+    }
+
+    const nonRootIndex = nodeHandle - this._rootCount;
+    const rangeStart =
+      this._invertedNonRootCallNodeTable.suffixOrderIndexRangeStart[
+        nonRootIndex
+      ];
+    const rangeEnd =
+      this._invertedNonRootCallNodeTable.suffixOrderIndexRangeEnd[nonRootIndex];
+    return [rangeStart, rangeEnd];
+  }
+
+  /**
+   * Materialize inverted call nodes for parentNodeHandle's children in the
+   * inverted tree.
+   *
+   * The returned array of call node handles is sorted by func.
+   */
+  _createChildren(
+    parentNodeHandle: InvertedCallNodeHandle
+  ): InvertedCallNodeHandle[] {
+    const parentDeepNodes =
+      this._takeDeepNodesForInvertedNode(parentNodeHandle);
+    const childrenInfo = this._computeChildrenInfo(
+      parentNodeHandle,
+      parentDeepNodes
     );
+    if (childrenInfo === null) {
+      // This node has no children.
+      return [];
+    }
+
+    return this._createChildrenForInfo(childrenInfo, parentNodeHandle);
+  }
+
+  /**
+   * Compute the information needed to create the children of parentNodeHandle.
+   *
+   * As we go deeper into the inverted tree, we go higher up in the non-inverted
+   * tree: To create the children of an inverted node, we need to look at the
+   * parents / "prefixes" of the corresponding non-inverted "deep nodes".
+   *
+   * See the class documentation for more details and examples.
+   */
+  _computeChildrenInfo(
+    parentNodeHandle: InvertedCallNodeHandle,
+    parentDeepNodes: Uint32Array
+  ): ChildrenInfo | null {
+    const parentDeepNodeCount = parentDeepNodes.length;
+    const [parentIndexRangeStart, parentIndexRangeEnd] =
+      this.getSuffixOrderIndexRangeForCallNode(parentNodeHandle);
+    if (parentIndexRangeStart + parentDeepNodeCount !== parentIndexRangeEnd) {
+      throw new Error('indexes out of sync');
+    }
+
+    const callNodeTable = this._callNodeTable;
+
+    // Count how many of the parent's deep nodes end at the parent. If there
+    // are any, they will all be at the start of the parentDeepNodes array by
+    // construction of the suffix order.
+    let nodesWhichEndHereCount = 0;
+    while (nodesWhichEndHereCount < parentDeepNodeCount) {
+      const deepNode = parentDeepNodes[nodesWhichEndHereCount];
+      if (callNodeTable.prefix[deepNode] !== -1) {
```

Comment:
nit: add a comment such as "This is the first deepNode that has a parent => all the following ones will have a parent as well."